python/ray/serve/_private/autoscaling_state.py

```diff
@@ -0,0 +1,355 @@
+            == num_replicas_running_at_target_version
+        )
+
+    def apply_bounds(self, num_replicas: int):
+        """Clips a replica count with current autoscaling bounds.
+
+        This takes into account target capacity.
+        """
+
+        return self.autoscaling_policy_manager.apply_bounds(
+            num_replicas,
+            self._target_capacity,
+            self._target_capacity_direction,
+        )
+
+    def record_request_metrics_for_replica(
+        self, replica_id: ReplicaID, window_avg: Optional[float], send_timestamp: float
+    ) -> None:
+        """Records average number of ongoing requests at a replica."""
+
+        if window_avg is None:
+            return
+
+        if (
+            replica_id not in self._replica_average_ongoing_requests
+            or send_timestamp > self._replica_average_ongoing_requests[replica_id][0]
+        ):
+            self._replica_average_ongoing_requests[replica_id] = (
+                send_timestamp,
+                window_avg,
+            )
+
+    def record_request_metrics_for_handle(
+        self,
+        *,
+        handle_id: str,
+        actor_id: Optional[str],
+        handle_source: DeploymentHandleSource,
+        queued_requests: float,
+        running_requests: Dict[ReplicaID, float],
+        send_timestamp: float,
+    ) -> None:
+        """Records average number of queued and running requests at a handle for this
+        deployment.
+        """
+
+        if (
+            handle_id not in self._handle_requests
+            or send_timestamp > self._handle_requests[handle_id].timestamp
+        ):
+            self._handle_requests[handle_id] = HandleRequestMetric(
+                actor_id=actor_id,
+                handle_source=handle_source,
+                queued_requests=queued_requests,
+                running_requests=running_requests,
+                timestamp=send_timestamp,
+            )
+
+    def drop_stale_handle_metrics(self, alive_serve_actor_ids: Set[str]) -> None:
+        """Drops handle metrics that are no longer valid.
+
+        This includes handles that live on Serve Proxy or replica actors
+        that have died AND handles from which the controller hasn't
+        received an update for too long.
+        """
+
+        timeout_s = max(
+            2 * self.autoscaling_policy_manager.get_metrics_interval_s(),
+            RAY_SERVE_MIN_HANDLE_METRICS_TIMEOUT_S,
+        )
+        for handle_id, handle_metric in list(self._handle_requests.items()):
+            # Drop metrics for handles that are on Serve proxy/replica
+            # actors that have died
+            if (
+                handle_metric.is_serve_component_source
+                and handle_metric.actor_id not in alive_serve_actor_ids
+            ):
+                del self._handle_requests[handle_id]
+                if handle_metric.total_requests > 0:
+                    logger.debug(
```

Comment:
This is an "expected" path, since replicas/proxies can be removed over time and we want to clear the handle metrics from them when that happens, so I think it should probably be debug to avoid spamming the controller logs.